naga/src/lib.rs

```diff
@@ -1284,6 +1289,46 @@ pub enum SwizzleComponent {
     W = 3,
 }
 
+#[derive(Clone, Copy, Debug, Hash, Eq, Ord, PartialEq, PartialOrd)]
+#[cfg_attr(feature = "serialize", derive(Serialize))]
+#[cfg_attr(feature = "deserialize", derive(Deserialize))]
+#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
+pub enum GatherMode {
+    BroadcastFirst,
+    Broadcast(Handle<Expression>),
+    Shuffle(Handle<Expression>),
+    ShuffleDown(Handle<Expression>),
+    ShuffleUp(Handle<Expression>),
+    ShuffleXor(Handle<Expression>),
+}
+
+#[derive(Clone, Copy, Debug, Hash, Eq, Ord, PartialEq, PartialOrd)]
+#[cfg_attr(feature = "serialize", derive(Serialize))]
+#[cfg_attr(feature = "deserialize", derive(Deserialize))]
+#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
+pub enum SubgroupOperation {
+    All = 0,
+    Any = 1,
```

Comment:
`And`, `Or`, `Xor` are not arithmetic either.
But they are all paired with a `CollectiveOperation`.

In fact:
- `All` is the same as `And` with `CollectiveOperation::Reduce` and scalar boolean operands.
- `Any` is the same as `Or` with `CollectiveOperation::Reduce` and scalar boolean operands.

`Elect` is does not fit at all as it takes no operands. It is more like `GatherMode::BroadcastFirst` as it can be emulated as `broadcastFirst(subgroup_invocation_id) == subgroup_invocation_id`.